pennylane/ops/qubit.py

```diff
@@ -2961,6 +2961,73 @@ def diagonalizing_gates(self):
         return [QubitUnitary(self.eigendecomposition["eigvec"].conj().T, wires=list(self.wires))]
 
 
+class Projector(Observable):
+    r"""Projector(basis_state, wires)
+    The basis state observable :math:`P=\ket{i}\bra{i}`.
+
+    The expectation of this observable returns the value
+
+    .. math::
+        \braket{P} = |\braketT{\psi}{\i}|^{2}
+
+    corresponding to the probability of measuring the quantum state in the :math:`i` -th eigenstate.
+
+    **Details:**
+
+    * Number of wires: Any
+    * Number of parameters: 1
+    * Gradient recipe: None
+
+    Args:
+        basis_state (tensor-like): binary input of shape ``(n, )``
+        wires (Iterable): wires that the template acts on
+    """
+    num_wires = AnyWires
+    num_params = 1
+    par_domain = "A"
+
+    def __init__(self, basis_state, wires, do_queue=True):
+        wires = Wires(wires)
+        shape = qml.math.shape(basis_state)
+
+        if len(shape) != 1:
+            raise ValueError(f"Basis state must be one-dimensional; got shape {shape}.")
+
+        n_basis_state = shape[0]
+        if n_basis_state != len(wires):
+            raise ValueError(
+                f"Basis state must be of length {len(wires)}; got length {n_basis_state}."
+            )
+
+        basis_state = list(qml.math.toarray(basis_state))
+
+        if not set(basis_state).issubset({0, 1}):
+            raise ValueError(f"Basis state must only consist of 0s and 1s; got {basis_state}")
+
+        super().__init__(basis_state, wires=wires, do_queue=do_queue)
+
+    @classmethod
+    def _eigvals(cls, *params):
+        """Eigenvalues of the specific projector operator.
+
+        Returns:
+            array: eigenvalues of the projector observable in the computational basis
+        """
+        w = np.zeros(2 ** len(params[0]))
+        idx = int("".join(str(i) for i in params[0]), 2)
```

Comment:
An alternative here could be
```suggestion
        idx = np.ravel_multi_index(params[0], [2] * len(params[0]))
```

Since this is purely using NumPy (and avoiding string manipulation), it might be more performant.

Reply:
Although, I might be mistaken. Here are some rough benchmarks I just ran:

```pycon
>>> basis_state = [1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 1, 0, 1]
>>> %timeit np.ravel_multi_index(basis_state, [2] * len(basis_state))
13.9 µs ± 516 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
>>> %timeit int("".join(str(i) for i in basis_state), 2)
8.33 µs ± 145 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

Reply:
thanks a lot for checking, @josh146 ! Also, I just checked that the ```ravel_multi_index``` function cannot handle states with 32 or more qubits.
```
>>> basis_state = np.random.randint(2, size=(32))
>>> np.ravel_multi_index(basis_state, [2] * len(basis_state))
ValueError: too many dimensions passed to ravel_multi_index
```
I will keep the original code in.

Reply:
Perfect, thanks @wongwsvincent!